pallets/liquidity-pools-forwarder/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+	pub trait Config: frame_system::Config {
+		/// Required origin for configuring domain forwarding.
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The Liquidity Pools message type.
+		type Message: LpMessageT<Domain = Domain>
+			+ Clone
+			+ Debug
+			+ PartialEq
+			+ Eq
+			+ MaxEncodedLen
+			+ TypeInfo
+			+ FullCodec;
+
+		/// The entity of the messages coming from this chain.
+		type MessageSender: MessageSender<
+			Middleware = Self::RouterId,
+			Origin = DomainAddress,
+			Message = Self::Message,
+		>;
+
+		/// The entity which acts on unwrapped messages.
+		type MessageReceiver: MessageReceiver<
+			Middleware = Self::RouterId,
+			Origin = DomainAddress,
+			Message = Self::Message,
+		>;
+
+		/// An identification of a router.
+		type RouterId: Parameter + MaxEncodedLen;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Forwarding info was set
+		ForwarderSet {
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		},
+		/// Forwarding info was removed
+		ForwarderRemoved {
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		},
+	}
+
+	/// Maps a router id to its forwarding info.
+	///
+	/// Can only be mutated via admin origin.
+	#[pallet::storage]
+	pub type RouterForwarding<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::RouterId, ForwardInfo, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The router id does not have any forwarder info stored
+		ForwardInfoNotFound,
+		/// Failed to unwrap a message which should be a forwarded one
+		UnwrappingFailed,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(0)]
+		pub fn set_forwarder(
+			origin: OriginFor<T>,
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			RouterForwarding::<T>::insert(
+				&router_id,
+				ForwardInfo {
+					source_domain,
+					contract: forwarding_contract,
+				},
+			);
+
+			Self::deposit_event(Event::<T>::ForwarderSet {
+				router_id,
+				source_domain,
+				forwarding_contract,
+			});
+
+			Ok(())
+		}
+
+		/// Remove the forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(1)]
+		pub fn remove_forwarder(origin: OriginFor<T>, router_id: T::RouterId) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			if let Some(info) = RouterForwarding::<T>::take(&router_id) {
```

Comment:
Nit - for other extrinsics that perform removal, we return a `NotFound` error, it might make sense here as well to inform the user of a potential incorrect call.

Reply:
Implemented, thanks!